File: internal/tool/edit_file.go
package tool

import (
	"context"
	"fmt"
	"os"
	"strings"
	"unicode/utf8"

	"github.com/TmkShink/nash-agent/internal/core"
	"github.com/TmkShink/nash-agent/internal/workspace"
)

type EditFile struct {
	workspace *workspace.Workspace
}

func NewEditFile(workspace *workspace.Workspace) *EditFile {
	return &EditFile{workspace: workspace}
}

func (t *EditFile) Definition() core.ToolDefinition {
	return core.ToolDefinition{
		Name:        "edit_file",
		Description: "Replace one exact, unique text fragment in an existing workspace file. The edit fails when the old text is missing or ambiguous.",
		InputSchema: objectSchema(map[string]any{
			"path":     map[string]any{"type": "string", "description": "Workspace-relative file path"},
			"old_text": map[string]any{"type": "string", "description": "Exact text that must occur once"},
			"new_text": map[string]any{"type": "string", "description": "Replacement text"},
		}, "path", "old_text", "new_text"),
	}
}

func (*EditFile) Effect() Effect { return EffectWrite }

func (*EditFile) Preview(arguments string) string {
	var args struct {
		Path    string `json:"path"`
		OldText string `json:"old_text"`
		NewText string `json:"new_text"`
	}
	if err := DecodeArguments(arguments, &args); err != nil {
		return "edit file"
	}
	return fmt.Sprintf("edit %s (-%d +%d bytes)", args.Path, len(args.OldText), len(args.NewText))
}

func (t *EditFile) Execute(ctx context.Context, arguments string) Result {
	if err := ctx.Err(); err != nil {
		return Failure("edit cancelled: "+err.Error(), map[string]any{"kind": "cancelled"})
	}
	var args struct {
		Path    string `json:"path"`
		OldText string `json:"old_text"`
		NewText string `json:"new_text"`
	}
	if err := DecodeArguments(arguments, &args); err != nil {
		return Failure("invalid edit_file arguments: "+err.Error(), map[string]any{"kind": "invalid_arguments"})
	}
	if args.OldText == "" {
		return Failure("old_text must not be empty", map[string]any{"kind": "invalid_arguments"})
	}
	if len(args.OldText)+len(args.NewText) > maxFileWriteBytes {
		return Failure("edit text exceeds the size limit", map[string]any{"kind": "limit_exceeded"})
	}

	path, err := t.workspace.ResolveExisting(args.Path)
	if err != nil {
		return Failure("cannot resolve file: "+err.Error(), map[string]any{"kind": "path_error"})
	}
	info, err := os.Stat(path)
	if err != nil {
		return Failure("cannot stat file: "+err.Error(), map[string]any{"kind": "io_error"})
	}
	if !info.Mode().IsRegular() {
		return Failure("path is not a regular file", map[string]any{"kind": "invalid_file_type"})
	}
	if info.Size() > int64(maxFileWriteBytes) {
		return Failure("file exceeds the editable size limit", map[string]any{"kind": "limit_exceeded"})
	}
	content, err := os.ReadFile(path)
	if err != nil {
		return Failure("cannot read file: "+err.Error(), map[string]any{"kind": "io_error"})
	}
	if !utf8.Valid(content) {
		return Failure("file is not valid UTF-8", map[string]any{"kind": "invalid_encoding"})
	}
	count := strings.Count(string(content), args.OldText)
	if count == 0 {
		return Failure("old_text was not found; read the file again before editing", map[string]any{"kind": "stale_edit"})
	}
	if count > 1 {
		return Failure(fmt.Sprintf("old_text occurs %d times; include more context", count), map[string]any{"kind": "ambiguous_edit", "matches": count})
	}

	updated := strings.Replace(string(content), args.OldText, args.NewText, 1)
	if len(updated) > maxFileWriteBytes {
		return Failure("edited file exceeds the size limit", map[string]any{"kind": "limit_exceeded"})
	}
	if err := ctx.Err(); err != nil {
		return Failure("edit cancelled: "+err.Error(), map[string]any{"kind": "cancelled"})
	}
	if err := writeFileAtomic(path, []byte(updated), info.Mode(), true); err != nil {
		return Failure("cannot edit file: "+err.Error(), map[string]any{"kind": "io_error"})
	}
	return Success(fmt.Sprintf("updated %s", t.workspace.Relative(path)), map[string]any{
		"path":         t.workspace.Relative(path),
		"bytes_before": len(content),
		"bytes_after":  len(updated),
	})
}

File: src/tools/edit-file.ts
import { readFile, stat } from "node:fs/promises";

import type { ToolDefinition } from "../core/types.js";
import { safeDisplay } from "../text/safe-display.js";
import type { Workspace } from "../workspace/workspace.js";
import { objectSchema, parseArgumentObject, requiredString } from "./arguments.js";
import {
  MAX_FILE_WRITE_BYTES,
  InvalidEncodingError,
  decodeUtf8,
  messageOf,
  writeFileAtomic,
} from "./file-helpers.js";
import {
  type LocalTool,
  type PreparedToolCall,
  ToolInputError,
  failure,
  success,
} from "./types.js";

export class EditFileTool implements LocalTool {
  public readonly effect = "write" as const;
  public readonly definition: ToolDefinition = {
    name: "edit_file",
    description:
      "Replace one exact, unique text fragment in an existing workspace file. Missing or ambiguous context is rejected.",
    parameters: objectSchema(
      {
        path: { type: "string", description: "Workspace-relative file path" },
        old_text: { type: "string", description: "Exact text that must occur once" },
        new_text: { type: "string", description: "Replacement text" },
      },
      ["path", "old_text", "new_text"],
    ),
  };

  readonly #workspace: Workspace;

  public constructor(workspace: Workspace) {
    this.#workspace = workspace;
  }

  public prepare(argumentsJson: string): PreparedToolCall {
    const object = parseArgumentObject(argumentsJson, ["path", "old_text", "new_text"]);
    const requestedPath = requiredString(object, "path", { nonBlank: true });
    const oldText = requiredString(object, "old_text");
    const newText = requiredString(object, "new_text");
    if (oldText === "") {
      throw new ToolInputError("old_text must not be empty");
    }
    if (
      Buffer.byteLength(oldText) > MAX_FILE_WRITE_BYTES ||
      Buffer.byteLength(newText) > MAX_FILE_WRITE_BYTES
    ) {
      throw new ToolInputError("edit text exceeds the size limit", "limit_exceeded");
    }
    return {
      preview: `edit ${safeDisplay(requestedPath)} (-${Buffer.byteLength(oldText)} +${Buffer.byteLength(newText)} bytes)`,
      execute: async (signal) =>
        await this.#execute(requestedPath, oldText, newText, signal),
    };
  }

  async #execute(
    requestedPath: string,
    oldText: string,
    newText: string,
    signal: AbortSignal,
  ) {
    if (signal.aborted) {
      return failure("edit was cancelled", "cancelled");
    }
    let resolved: string;
    try {
      resolved = await this.#workspace.resolveExisting(requestedPath);
    } catch (error) {
      return failure(`cannot resolve file: ${messageOf(error)}`, "path_error");
    }

    try {
      const information = await stat(resolved);
      if (!information.isFile()) {
        return failure("path is not a regular file", "invalid_file_type");
      }
      if (information.size > MAX_FILE_WRITE_BYTES) {
        return failure("file exceeds the editable size limit", "limit_exceeded");
      }
      const bytes = await readFile(resolved);
      if (bytes.byteLength > MAX_FILE_WRITE_BYTES) {
        return failure("file exceeds the editable size limit", "limit_exceeded");
      }
      if (bytes.includes(0)) {
        return failure("file appears to be binary", "binary_file");
      }
      const content = decodeUtf8(bytes);
      const occurrences = countOccurrences(content, oldText);
      if (occurrences === 0) {
        return failure(
          "old_text was not found; read the file again before editing",
          "stale_edit",
        );
      }
      if (occurrences > 1) {
        return failure(
          `old_text occurs ${occurrences} times; include more context`,
          "ambiguous_edit",
          { matches: occurrences },
        );
      }

      const updated = content.replace(oldText, newText);
      const updatedBytes = Buffer.byteLength(updated);
      if (updatedBytes > MAX_FILE_WRITE_BYTES) {
        return failure("edited file exceeds the size limit", "limit_exceeded");
      }
      if (signal.aborted) {
        return failure("edit was cancelled", "cancelled");
      }
      await writeFileAtomic(resolved, updated, information.mode & 0o777, true);
      return success(`updated ${this.#workspace.relative(resolved)}`, {
        path: this.#workspace.relative(resolved),
        bytes_before: bytes.byteLength,
        bytes_after: updatedBytes,
      });
    } catch (error) {
      if (error instanceof InvalidEncodingError) {
        return failure(error.message, "invalid_encoding");
      }
      return failure(`cannot edit file: ${messageOf(error)}`, "io_error");
    }
  }
}

function countOccurrences(content: string, fragment: string): number {
  let count = 0;
  let offset = 0;
  while (true) {
    const index = content.indexOf(fragment, offset);
    if (index === -1) {
      return count;
    }
    count++;
    offset = index + fragment.length;
  }
}
